Add Key Rotation to MessageEncryptor and MessageVerifier and simplify the Cookies middleware

mjc-gh opened the pull request:
### Summary

This PR introduces `ActiveSupport::KeyRotator` which wraps `KeyGenerator` as well as `MessageEncryptor` and `MessageVerifier` and provides an interface for easily rotating between different encryption ciphers or message digests, salts, and secrets.

This PR also simplifies the cookies middleware and removes several of the internal legacy middlewares. Upgrading both signed and encrypted legacy cookies is now handled by rotation capable middlewares which each use a `KeyRotator` instance. Is also introduces a new interface for configuring the middleware.

### Other Information

This feature spawned from discussions had in the Rails’ basecamp.

I still need to update the configuring and security markdown guides with details of this change. I plan on adding a new section the security guide detailing how secrets and salts can be rotated. I also want to add more rotation tests to `railties/test/application/middleware/cookies_test.rb`.

Any and all feedback is welcomed!

/cc @matthewd @kaspth

bdewater commented:
Took a brief look at the code, looks nice 👏 One question: the rotator in it's current state seems to be for upgrading cookies and people using MessageEncryptor defaults, which is probably what most people need anyway. But passing in symbols and raising if it doesn't match a known scheme excludes people who use MessageEncryptor now with their own settings.

The shorthands for old and new defaults make sense, but it isn't is possible to pass in your own legacy scheme hash that is just passed on to create a new MessageEncryptor. Likewise "only `:aes_gcm` may be used for the primary scheme" seems very prescriptive and not in line with the 'provide sharp knives' philosophy. Should we make it a bit more general so it can rotate all configurations of MessageEncryptor?

kaspth commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -13,12 +13,12 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_dispatch.perform_deep_munge = true
 
+    config.action_dispatch.legacy_encryptor_schemes = []
+    config.action_dispatch.legacy_verifier_schemes  = []
```
I find the `scheme` wording confusing. Isn't that what `aes-256-gcm` is? Also dislike the `legacy` wording since it doesn't clarify how people can use this to rotate their keys.

Ideally I'd like the config to be something like:

```ruby
# config/initializers/rotations.rb
# Here's some instructions on how to do cookie rotations right and why should be.
#   - Rotate your secret key base whenever a coworker leaves, etc.
Rails.application.config.secrets.tap do |secrets|
  secrets.rotate_out secret_key_base: secrets.old_secret_key_base
  secrets.rotate_out cipher: 'aes-256-gcm'
  secrets.rotate_out :signed, digest: 'xxx' # We need some way to explain that a rotation targets either signed or encrypted messages.
end

Rails.application.config.action_dispatch.tap do |config|
  config.cookies.rotate_out signed_salt: config.old_signed_salt # It'll use the new config.signed_salt but add this as a fallback.
end
```

With that API our new framework defaults could then contain:

```ruby
# config/initializers/new_framework_defaults_5_2.rb
# …

# Use AES 256 GCM authenticated encryption in cookies and ActiveSupport::MessageEncryptor.
# Rails.application.secrets.rotate_out cipher: 'aes-xx-cbc'
```

mikeycgto replied:
I agree with these wording changes.

The word "scheme" is unclear and could refer to several things really. It's probably best to just avoid that term altogether.

Removing the "legacy" wording is a good idea as well since this feature is more than just legacy support. Your configuration above is really nice and clean! I can rework my changes to introduce this style of configuration for this feature.

mikeycgto replied:
@kaspth any thoughts on what I should call a complete configuration for a cipher/digest, the salts, and a secret?

I'm thinking I'm just going to call it an `encryptor_config` and `verifiy_config` within the code but am open to something more fitting here!

kaspth replied:
How about just `verifiers` and `encryptors`? We should support passing a message verifier or encryptor instance to `rotate_out`, which I forgot to add above.

I'm still on the fence about the `_out` suffix. Do think you think just `config.secrets.rotate` would be clear enough to imply that it's the old value that goes there?

mikeycgto replied:
I like the idea of passing an encryptor or verifier directly. This would be for more advanced uses and would be great to support.

Under the hood, I think we can just have these instances be passed to `KeyRotator` and have the class able to insert arbitrary encryptors or verifiers into its chain. Modifying `KeyRotator` to support this should somewhat address @bdewater's feedback about making `KeyRotator` more versatile in general.

I also think the `config.secrets.rotate` syntax should be clear enough. I feel the "rotating out" concept is implied by the feature as a whole.

kaspth reviewed (commented):
Made a focused review on how I'd like to bring the idea of key rotation to the fore for users (rather than focusing purely on our need to upgrade their cookies) 😊

Appreciate the work you've been doing so far! You seem really comfortable in the codebase 😄

mikeycgto commented:
Thanks @bdewater and @kaspth for the feedback!

As far as `:cipher` options go for the `KeyRotator`, I think I'm going to drop the symbol usage and just use the same strings `MessageEncryptor` (and thus OpenSSL) would expect. For instance, instead of using `cipher: :aes_gcm` we can just use `cipher: 'aes-256-gcm'`. 

> One question: the rotator in it's current state seems to be for upgrading cookies and people using MessageEncryptor defaults, which is probably what most people need anyway. But passing in symbols and raising if it doesn't match a known scheme excludes people who use MessageEncryptor now with their own settings.

I agree with this overall, @bdewater. Opening up `KeyRotator` to work with potentially any cipher option for `MessageEncrytor` would be a really nice feature. What would the expectation here for using the `KeyRotator` without a defined preset? The tricky part here is whether or not the supplied `secret` string should be passed through a `KeyGenerator` or not.

Maybe, in general, we can leave the `:secret` option as. It's mostly meant to mirror `secret_key_base` and should be used with a `KeyGenerator` and salt for deriving actual keys. Maybe we can then add a alternative `:key` option to the configuration which is _not_ run through a `KeyGenerator` is passed directly to either `MessageEncryptor` or `MessageVerifier`. This sort of option would be meant for more advanced users who are generating sufficiently random and long keys on their own. The `:key` option would then not require a `:salt` option since no key derivation is occurring.

I'm going to rework this class some more such that all potential uses are easily supported in a clear and understandable API. It's far to versatile of a feature to leave for just preset options. Thanks very much for your feedback here! 

> Likewise "only :aes_gcm may be used for the primary scheme" seems very prescriptive and not in line with the 'provide sharp knives' philosophy.

This comment is actually out-of-date now. Prior to submitting this PR I had limits in place with what could be used as the `primary_scheme`. My original thinking here is that security APIs should be very easy to use correctly. Maybe we can explore this aspect a bit elsewhere with warnings or something...?

bdewater commented:
> Maybe, in general, we can leave the :secret option as. It's mostly meant to mirror secret_key_base and should be used with a KeyGenerator and salt for deriving actual keys. Maybe we can then add a alternative :key option to the configuration which is not run through a KeyGenerator is passed directly to either MessageEncryptor or MessageVerifier. This sort of option would be meant for more advanced users who are generating sufficiently random and long keys on their own. The :key option would then not require a :salt option since no key derivation is occurring.

This sounds like a great approach! Better defaults while still allowing a configuration to fully preserve existing behaviour for interoperability 👍  My worry with combining MessageEncryptor with secret_key_base is that changing it could break more than just existing sessions, e.g. when you're using it to exchange encrypted messages across different (web/mobile/etc) apps.

bdewater commented:
Also a heads-up: you'll want to add the frozen string magic comment to new files, see https://github.com/rails/rails/pull/29728 :)

rafaelfranca commented on railties/lib/rails/application.rb:
```diff
@@ -257,13 +257,15 @@ def env_config
-          "action_dispatch.cookies_digest" => config.action_dispatch.cookies_digest
-        )
+        ).tap do |config|
+          ActionDispatch::CookieSecurity.configure! config
```
The Railtie component can't use ActionDispatch constants. We should inverse the dependency.

mikeycgto replied:
I had a feeling this was the case. I'm going to rework how the various configuration defaults are set, fix the dependency issues, and remove the top-level `CookieSecurity` constant.

rafaelfranca commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -37,7 +37,7 @@ class Railtie < Rails::Railtie # :nodoc:
       ActionDispatch::ExceptionWrapper.rescue_templates.merge!(config.action_dispatch.rescue_templates)
 
-      config.action_dispatch.authenticated_encrypted_cookie_salt = "authenticated encrypted cookie" if config.action_dispatch.use_authenticated_cookie_encryption
+      ActionDispatch::CookieSecurity.configure! config.action_dispatch
```
This is a normal hash, so how `config.action_dispatch["action_dispatch.secret_key_base"]` would return something?

rafaelfranca commented on actionpack/lib/action_dispatch/cookie_security.rb:
```diff
@@ -0,0 +1,68 @@
+module ActionDispatch
+  module CookieSecurity
```
I'm not a fan of this top level constant. This class seems to be private API so maybe move to inside the `actionpack/lib/action_dispatch/middleware/cookies.rb ` file?

rafaelfranca commented on activesupport/lib/active_support/key_rotator.rb:
```diff
@@ -0,0 +1,210 @@
+
+    delegate :encrypt_and_sign, :decrypt_and_verify, to: :@primary_encryptor
+
+    def decrypt_and_verify_with_legacy_encryptors(value)
```
Should not the key rotator wrap the legacy logic automatically? Like, why would someone use the key rotator and use `decrypt_and_verify` without rescuing the exception and calling `decrypt_and_verify_with_legacy_encryptors`?

mikeycgto replied:
This is how I originally designed the `KeyRotator` class. The problem was, in the cookie middleware `parse` methods, we need to know when the primary encryptor/verifier fails so that the cookie can be updated.

I experimented with having `decrypt_and_verify` that accepts an optional block that is called when the primary encryptor/verifier fails. It didn't do the trick though because you end up having to `deserialize` before doing the actual cooke update.

Definitely open to ideas on how we can have `decrypt_and_verify` and `verify` signal to their caller that primary encryptor or verifier were not used. This would definitely result in a simpler and easier to use API as you have suggested.

rafaelfranca replied:
maybe we could invert that. Implement a `decrypt_and_verify_without_legacy_encryptors` (or just exposing the `primary_encryptor` for those who want control when the primary encryptor could not be used.

kaspth commented:
I think using both `key` and `secret` wording seems confusing. How would users know which has passed through a `key_generator` and which has not?

I'm having trouble remembering right now for instance 😅

Perhaps it could be `generated_key`?

mikeycgto commented:
> I think using both key and secret wording seems confusing. How would users know which has passed through a key_generator and which has not?
> 
> I'm having trouble remembering right now for instance :sweat_smile:
> 
> Perhaps it could be generated_key?

Maybe `raw_key` or even `raw_key_material` is a better name for what it actually is?

This option will also be clearly documented in that it is used as an actual cipher key and should be sufficiently long, random, and properly encoded. That is, it should probably be a raw byte string from `SecureRandom` or something similar. It should *not* be encoded as a hex string or otherwise either, which can potentially lead to losses in entropy.

mikeycgto commented:
With some great advice from @kaspth, I reworked this PR and added the rotation features directly to the `MessageEncryptor` and `MessageVerifier` classes with the help of some small mixins. Overall this approach feels a lot better and it addresses a lot of the feedback raised in this PR.

There are still some failing tests with regards to "upgrading" `secret_token` HMAC cookies to encrypted cookies. I should be able to easily handle this case in the cookies middleware directly with a legacy verifier in the `EncryptedKeyRotatingCookieJar` middleware class. I also need to add some more docs and update the configuration guide with notes on several new cookie middleware configuration options. These options should give a high degree of flexibility to developers and will enable them to use different digests for signed cookies (like SHA256) and different ciphers for encrypted cookies.

As always, any and all feedback is welcomed!

rafaelfranca commented on actionpack/lib/action_dispatch.rb:
```diff
@@ -64,6 +64,7 @@ class IllegalStateError < StandardError
     autoload :Static
   end
 
+  autoload :CookieSecurity
```
Is this needed? I tried to find where this constant is defined and could not.

mikeycgto replied:
Yikes, this is stale code. I can remove it.

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -548,63 +561,92 @@ def serializer
       private
         def parse(name, signed_message)
-          deserialize name, @verifier.verified(signed_message)
+          deserialize(name, @verifier.verify(signed_message) { |message|
```
`do/end` here. If needed it is better to extract to a local variable than using `{` to change the precedence.

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -548,63 +561,92 @@ def serializer
         def parse(name, signed_message)
-          deserialize name, @verifier.verified(signed_message)
+          deserialize(name, @verifier.verify(signed_message) { |message|
+            return deserialize(name, message).tap { |deserialized| self[name] = { value: deserialized } }
```
It is not clear why you need to use `return` here. It seems it is because this block is called more than once. So maybe the method should be clear about that.

mikeycgto replied:
The `verify` method (and the `decrypt_and_verify` method) call an optional block when a rotated instance is used to verify the message. We need to rewrite the cookie's value when this occurs and this optional block is utilized to signal that.

The tricky part is we need actually write the deserialize message so within this block `deserialize` is called and the cookie is updated. The `return` is there so the `parse` method returns the deserialized value from within the block. If we didn't return here, the outer `deserialize` would also get invoked and deserialization would effectively occur twice.

I do agree that `return` here is a little confusing. I'm definitely open to a better way of signal to the caller of `verify` or `decrypt_and_verify` when a rotated instance is used so that we can better handle this case.

rafaelfranca commented on activesupport/lib/active_support/messages/rotator.rb:
```diff
@@ -0,0 +1,62 @@
+
+module ActiveSupport
+  module Messages
+    module Rotator # :nodoc:
```
This will make all the constants inside it to be not documented.

mikeycgto commented:
One last change I'm considering making to this PR is to provide the default salt values for both middleware's `expand_outbound_rotation` methods. This way, when defining rotation options, users do not have redefine the default salt values. Thoughts?

Note that, this is a little more involved than a simple default because of the different in salt values for different encryption schemes (CBC mode vs GCM cookies).

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -249,39 +258,42 @@ def signed_or_encrypted
+        end
+
+        def signed_cookie_salt
+          request.signed_cookie_salt || ""
```
Do we need to provide these empty string fallbacks?

mikeycgto replied:
Don't think its necessary to have the fallbacks. I can change this.

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -249,39 +258,42 @@ def signed_or_encrypted
         end
 
-        def upgrade_legacy_hmac_aes_cbc_cookies?
+        def rotate_out_legacy_aes_cbc_hmac_cookies?
```
Let's just leave these methods to say `upgrade`. That's a different job than the rest of the code interpreting it as meaning "apply rotation".

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -560,63 +573,98 @@ def serializer
+        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: signed_cookie_digest, serializer: SERIALIZER)
+
+        Array.wrap(signed_cookie_rotate_out).each do |outbound|
+          @verifier.rotate(*expand_outbound_rotation(outbound))
```
I think your `expand_outbound_rotation` is doing too many things. Can it be simplified to something like?

```ruby
Array.wrap(signed_cookie_rotate_out).each do |digest:, **rotation|
  @verifier.rotate(extract_key(rotation), digest: digest, serializer: SERIALIZER)
end

def extract_key(raw_key: nil, secret: nil, salt: nil)
  if raw_key
    raw_key
  else
    raise ArgumentError, "Missing secret and salt to generate key from" unless secret && salt

    ActiveSupport::KeyGenerator.new(secret, iterations: 1000).generate_key(salt)
  end
end
```

mikeycgto replied:
I dig it :+1:

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -625,39 +673,42 @@ def commit(options)
+            return key, sign_key, options
+          end
+
+          return key, options
```
Also think this is doing too much. How about:

```ruby
Array.wrap(encrypted_cookie_rotate_out).each do |rotation|
  config = rotation.slice(:cipher, :digest).merge(serializer: serializer)

  @encryptor.rotate(extract_signing_key(rotation), extract_key(rotation), config)
end

if rotate_out_legacy_aes_cbc_hmac_cookies?
  config = { cipher: "aes-256-cbc", digest: digest }
  
  @encryptor.rotate(
    extract_signing_key(config.merge(key_generator: key_generator, signed_salt: encrypted_signed_cookie_salt),
    extract_key(config.merge(key_generator: key_generator, salt: encrypted_cookie_salt)),
    config
  )
end

def extract_key(cipher:, raw_key: nil, key_generator: nil, secret: nil, salt: nil, **)
  if raw_key
    raw_key
  else
    raise ArgumentError, "Missing salt to generate encryption key from" unless salt

    key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
    key_generator.generate_key(salt, ActiveSupport::MessageEncryptor.key_len(cipher))
  end
end

def extract_signing_key(cipher:, raw_signed_key: nil, key_generator: nil, secret: nil, signed_salt: nil, **)
  if cipher =~ AES_CBC_CIPHER_REGEX && !(raw_signed_key || signed_salt)
    raise ArgumentError, "Missing signed_salt or raw_signed_key when using AES-CBC"
  end

  if raw_signed_key
    raw_signed_key
  else
    raise ArgumentError, "Missing signed_salt to generate sign key from" unless signed_salt

    key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
    key_generator.generate_key(signed_salt)
  end
end
```

mikeycgto replied:
Refactored this a bit more and added a couple more test cases to the cookie middleware for rotating encrypted cookies.

kaspth reviewed (commented):
All I had time to review today. I'll check back in later.

We need to find a different way to be notified of messages being verified/decrypted with legacy verifiers. Executing that block reads poorly 😊

mikeycgto commented:
Sounds good @kaspth!

Going to think about a different approach for updating cookies that verified/decrypted with a rotated instance. I agree the `return` within the block is hard to read and follow the execution flow.

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -628,39 +673,32 @@ def commit(options)
+          raise ArgumentError, "Missing signed_salt to generate sign key from" unless signed_salt
+
+          key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
+          key_generator.generate_key(signed_salt)
```
Think we should pull these key extractions into the `rotate` methods on the verifier and encryptor.

I worked a bit on a different version of the rotations, which also includes a different approach to what happens on legacy rotations. How does this sound?

```ruby
# frozen_string_literal: true

module ActiveSupport
  module Messages
    module Rotator # :nodoc:
      def initialize(*)
        super

        @rotations = []
        @on_legacy_rotation = nil
      end

      def on_legacy_rotation(on_rotation)
        old_on_legacy_rotation, @on_legacy_rotation = @on_legacy_rotation, on_rotation

        yield
      ensure
        @on_legacy_rotation = old_on_legacy_rotation
      end

      def rotate(*args)
        @rotations << make_rotation(*args)
      end

      private
        def key_generator_for(secret)
          ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
        end

        def legacy_rotation(message)
          @on_legacy_rotation.call(message) if @on_legacy_rotation
        end

        def in_rotations
          @rotations.find do |rotation|
            mark_legacy_rotation(yield(rotation)) rescue next
          end
        end

      module Encryptor
        include Rotator

        def decrypt_and_verify(*args)
          super
        rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature => error
          in_rotations { |encryptor| encryptor.decrypt_and_verify(*args) } || raise(error)
        end

        private
          def make_rotation(raw_key: nil, raw_signed_key: nil, **options)
            self.class.new \
              raw_key || extract_key(options),
              raw_signed_key || extract_signing_key(options),
              options.slice(:cipher, :digest, :serializer)
          end
        
          def extract_key(salt:, key_generator: nil, secret: nil, cipher: nil, **)
            key_generator ||= key_generator_for(secret)
            key_generator.generate_key(salt, self.class.key_len(cipher))
          end

          def extract_signing_key(key_generator: nil, signed_salt: nil, secret: nil, **)
            unless cipher =~ /\Aaes-\d+-cbc\z/i
              raise ArgumentError, "missing signed_salt to generate signing key from"
            end

            key_generator ||= key_generator_for(secret)
            key_generator.generate_key(signed_salt)
          end
      end

      module Verifier
        include Rotator

        def verify(*args, &block)
          super
        rescue MessageVerifier::InvalidSignature => error
          in_rotations { |verifier| verifier.verify(*args) } || raise(error)
        end
      end

      private
        def make_rotation(raw_key: nil, digest: nil, serializer: nil, **options)
          self.class.new(raw_key || extract_key(options), digest: digest, serializer: serializer))
        end
      
        def extract_key(key_generator: nil, secret: nil, salt:)
          key_generator ||= key_generator_for(secret)
          key_generator.generate_key(salt)
        end
    end
  end
end
```

mikeycgto replied:
I agree we should move the key and options extraction to within the rotator. It's more useful that way.

Some questions about this code:
1. Where is `mark_legacy_rotation` defined? Should this be a call to `legacy_rotation` instead?
2. Using `on_legacy_rotation` to signal to the cookies middleware that a rotated instance was using may lead to deserializing the cookie twice. This is an issue I ran into earlier with a similar, callback-based approach.

kaspth reviewed (commented):
What I had time for in this pass, I'll be making another one. Let me know what you think :)

mikeycgto commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -256,35 +265,38 @@ def upgrade_legacy_signed_cookies?
+        end
+
+        def signed_cookie_salt
+          request.signed_cookie_salt
```
Agreed :+1:

mikeycgto commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -563,63 +576,95 @@ def serializer
+          return raw_key if raw_key
+          raise ArgumentError, "Missing secret and salt to generate key from" unless secret && salt
+
+          ActiveSupport::KeyGenerator.new(secret, iterations: 1000).generate_key(salt)
```
You mean accept an `:key_generator` option which would be a KeyGenerator instance?

mikeycgto commented:
Added more tests around the key option handling, which was just moved to the `ActiveSupport::Messages::Rotator` module. Definitely a lot better to integrate this code at this level since it will make `MessageEncrytor`/`MessageVerifier` more useful in general.

Refactored the cookies middleware to use these changes as well. Not sure I totally understand how you intended the `on_legacy_rotation` method to be used, so let me know what you think about changes to the two middleware's `parse` methods. I changed `on_legacy_rotation` to `on_rotation_used` since I think we should avoid using the word "legacy" as it has a lot of other potential meanings.

kaspth commented:
Yeah, that's what I had in mind for the `on_rotation_used` thing. Though I'm not too fond of the way it ended up looking.

I've worked on another alternative here by adding an `on_rotation` option. That'll then be called when the verifier/encryptor has resorted to rotating. (I realized it shouldn't take a message since it's just a notifier that rotation has taken place.)

It could prove dangerous in case somebody has subclassed verifier to override `verified`/`verify`, same for the encryptor, because there's now an extra keyword argument. But I don't know if people would be doing that.

I've also noticed that we're rotating on `verify` when we should hook in at the `verified` level instead. Then we'll support rotation for both.

To your worry about deserializing twice, I figured "hey, why not just let deserialize control when we've been rotating". It already handled upgrading for the serializer change, so it makes the best spot for this upgrading as well.
`deserialize` is a bit more bloated now, but the `case` statement really clears up the different code paths.

What do you think? 😊

```patch
diff --git a/actionpack/lib/action_dispatch/middleware/cookies.rb b/actionpack/lib/action_dispatch/middleware/cookies.rb
index a752329497..2b8d8862dd 100644
--- a/actionpack/lib/action_dispatch/middleware/cookies.rb
+++ b/actionpack/lib/action_dispatch/middleware/cookies.rb
@@ -535,12 +535,16 @@ def serialize(value)
           serializer.dump(value)
         end
 
-        def deserialize(name, value)
+        def deserialize(name)
+          rotate = false
+          value  = yield -> { rotate = true }
+
           if value
-            if needs_migration?(value)
-              Marshal.load(value).tap do |v|
-                self[name] = { value: v }
-              end
+            case
+            when needs_migration?(value)
+              self[name] = Marshal.load(value)
+            when rotate
+              self[name] = serializer.load(value)
             else
               serializer.load(value)
             end
@@ -584,16 +588,9 @@ def initialize(parent_jar)
 
       private
         def parse(name, signed_message)
-          write_upgraded_cookie = proc do |message|
-            return deserialize(name, message).tap { |value| self[name] = { value: value } }
-          end
-
-          @verifier.on_rotation_used(write_upgraded_cookie) do
-            deserialize(name, @verifier.verify(signed_message))
+          deserialize(name) do |rotate|
+            @verifier.verified(signed_message, on_rotation: rotate)
           end
-
-        rescue ActiveSupport::MessageVerifier::InvalidSignature
-          nil
         end
 
         def commit(options)
@@ -632,18 +629,11 @@ def initialize(parent_jar)
 
       private
         def parse(name, encrypted_message)
-          write_upgraded_cookie = proc do |message|
-            return deserialize(name, message).tap { |value| self[name] = { value: value } }
-          end
-
-          @encryptor.on_rotation_used(write_upgraded_cookie) do
-            deserialize(name, @encryptor.decrypt_and_verify(encrypted_message))
+          deserialize(name) do |rotate|
+            @encryptor.decrypt_and_verify(encrypted_message, on_rotation: rotate)
           end
-
         rescue ActiveSupport::MessageEncryptor::InvalidMessage, ActiveSupport::MessageVerifier::InvalidSignature
-          return try_legacy_verifier name, encrypted_message if defined? @legacy_verifier
-
-          nil
+          parse_legacy_signed_message(name, encrypted_message)
         end
 
         def commit(options)
@@ -652,13 +642,14 @@ def commit(options)
           raise CookieOverflow if options[:value].bytesize > MAX_COOKIE_SIZE
         end
 
-        def try_legacy_verifier(name, encrypted_message)
-          deserialize(name, @legacy_verifier.verify(encrypted_message)).tap do |deserialize|
-            self[name] = deserialize
-          end
+        def parse_legacy_signed_message(name, legacy_signed_message)
+          if defined?(@legacy_verifier)
+            deserialize(name) do |rotate|
+              rotate.call
 
-        rescue ActiveSupport::MessageVerifier::InvalidSignature
-          nil
+              @legacy_verifier.verified(legacy_signed_message)
+            end
+          end
         end
     end
 
diff --git a/activesupport/lib/active_support/message_encryptor.rb b/activesupport/lib/active_support/message_encryptor.rb
index aae3aab706..2018233f56 100644
--- a/activesupport/lib/active_support/message_encryptor.rb
+++ b/activesupport/lib/active_support/message_encryptor.rb
@@ -138,7 +138,7 @@ def encrypt_and_sign(value, expires_at: nil, expires_in: nil, purpose: nil)
 
     # Decrypt and verify a message. We need to verify the message in order to
     # avoid padding attacks. Reference: https://www.limited-entropy.com/padding-oracle-attacks/.
-    def decrypt_and_verify(data, purpose: nil)
+    def decrypt_and_verify(data, purpose: nil, **)
       _decrypt(verifier.verify(data), purpose)
     end
 
diff --git a/activesupport/lib/active_support/message_verifier.rb b/activesupport/lib/active_support/message_verifier.rb
index 36d85b9500..6aa4d9e2b9 100644
--- a/activesupport/lib/active_support/message_verifier.rb
+++ b/activesupport/lib/active_support/message_verifier.rb
@@ -133,7 +133,7 @@ def valid_message?(signed_message)
     #
     #   incompatible_message = "test--dad7b06c94abba8d46a15fafaef56c327665d5ff"
     #   verifier.verified(incompatible_message) # => TypeError: incompatible marshal file format
-    def verified(signed_message, purpose: nil)
+    def verified(signed_message, purpose: nil, **)
       if valid_message?(signed_message)
         begin
           data = signed_message.split("--".freeze)[0]
@@ -158,8 +158,8 @@ def verified(signed_message, purpose: nil)
     #
     #   other_verifier = ActiveSupport::MessageVerifier.new 'd1ff3r3nt-s3Krit'
     #   other_verifier.verify(signed_message) # => ActiveSupport::MessageVerifier::InvalidSignature
-    def verify(signed_message, purpose: nil)
-      verified(signed_message, purpose: purpose) || raise(InvalidSignature)
+    def verify(*args)
+      verified(*args) || raise(InvalidSignature)
     end
 
     # Generates a signed message for the provided value.
diff --git a/activesupport/lib/active_support/messages/rotator.rb b/activesupport/lib/active_support/messages/rotator.rb
index a1f7ab891f..14e9ed05ce 100644
--- a/activesupport/lib/active_support/messages/rotator.rb
+++ b/activesupport/lib/active_support/messages/rotator.rb
@@ -7,39 +7,24 @@ def initialize(*args)
         super
 
         @rotations = []
-        @on_rotation_used = nil
       end
 
       def rotate(*args)
         @rotations << create_rotation(*args)
       end
 
-      def on_rotation_used(on_rotation)
-        old_on_rotation_used, @on_rotation_used = @on_rotation_used, on_rotation
-
-        yield
-      ensure
-        @on_rotation_used = old_on_rotation_used
-      end
-
-      def has_rotations?
-        !@rotations.empty?
-      end
-
       module Encryptor
         include Rotator
 
         AES_CBC_EXPRESSION = /\Aaes-\d+-cbc\z/i.freeze
 
-        def decrypt_and_verify(*args)
+        def decrypt_and_verify(*args, on_rotation: nil)
           super
-
-        rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature => error
-          find_first_rotation { |encryptor| encryptor.decrypt_and_verify(*args) } || raise(error)
+        rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature
+          run_rotations(on_rotation) { |encryptor| encryptor.decrypt_and_verify(*args) } || raise
         end
 
         private
-
           def create_rotation(raw_key: nil, raw_signed_key: nil, **options)
             self.class.new \
               raw_key || extract_key(options),
@@ -65,15 +50,15 @@ def extract_signing_key(cipher:, signed_salt: nil, key_generator: nil, secret: n
       module Verifier
         include Rotator
 
-        def verify(*args)
-          super
-
-        rescue MessageVerifier::InvalidSignature => error
-          find_first_rotation { |verifier| verifier.verify(*args) } || raise(error)
+        def verified(*args, on_rotation: nil)
+          if verified = super
+            verified
+          else
+            run_rotations(on_rotation) { |verifier| verifier.verified(*args) }
+          end
         end
 
         private
-
           def create_rotation(raw_key: nil, digest: nil, serializer: nil, **options)
             self.class.new(raw_key || extract_key(options), digest: digest, serializer: serializer)
           end
@@ -85,18 +70,16 @@ def extract_key(key_generator: nil, secret: nil, salt:)
       end
 
       private
-
         def key_generator_for(secret)
           ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
         end
 
-        def call_on_rotation_used(message)
-          @on_rotation_used.call(message) if @on_rotation_used
-        end
-
-        def find_first_rotation
+        def run_rotations(on_rotation)
           @rotations.find do |rotation|
-            return yield(rotation).tap { |message| call_on_rotation_used message } rescue next
+            if message = yield(rotation) rescue next
+              on_rotation.call if on_rotation
+              return message
+            end
           end
         end
     end
```

mikeycgto commented on guides/source/security.md:
```diff
@@ -126,6 +126,97 @@ Older versions of Rails use CookieStore, which uses `secret_token` instead of `s
+  # First rotation options...
+}, {
+  # Second rotation options...
+}]
```
Agreed it doesn't feel Railsy enough. I like this approach a lot!

kaspth replied:
Cool! There's still some unanswered questions around how to integrate this into Action Dispatch, e.g. the different namespaces `config.rotations.secrets` versus `config.action_dispatch.rotations.cookies`.

mikeycgto replied:
Yeah I was wondering about the different namespaces. I figured `config.action_dispatch.rotations.cookies` would be the namespace consumed by the cookie middlewares. Can take the various options defined in the configuration and setup the cookies middleware accordingly with the desired rotations. What should the `config.rotations.secrets` namespace be used for?

Additionally, this example uses `:secret_key_base` instead of `:secret`. Should we change the Rotator `:secret` option to be `:secret_key_base`? This is effectively what the option maps to as far as Rails cookies go.

kaspth replied:
Yeah, exactly. I made a comment about how that would look elsewhere.

`config.rotations.secrets` was to rotate `secret_key_base` app-wide. But that likely needs more thought. And needs to look into how `secret_key_base` is used throughout. Come to think of it, it might not even be wrapped in a verifier/encryptor. So… let's shelve that part for now.

I think my usage of `:secret_key_base` above was a confusion on my part. It seems we never pass the secret directly but always generate a secret using the app's `key_generator`.

mikeycgto replied:
I actually recently published a blog post about the secret key base here: https://medium.com/@mikeycgto/understanding-the-secret-key-base-in-ruby-on-rails-ce2f6f9968a1

As far as I can tell, there are 3 main uses:

1. Encrypted cookies
2. Signed cookies
3. The application message verifier, which has grown in use with ActiveStorage

This PR covers the first 2 cases. I wonder if we should try to cover the last case with these changes?

kaspth replied:
Sweet! Let's leave it to another PR, this touches enough parts already.

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -563,24 +566,31 @@ def serializer
+        secret = key_generator.generate_key(signed_cookie_salt)
+        @verifier = ActiveSupport::MessageVerifier.new(secret, digest: signed_cookie_digest, serializer: SERIALIZER)
+
+        Array.wrap(signed_cookie_rotate_out).each do |rotation_options|
```
With the proposed `config.action_dispatch.rotations.cookies` this would probably read like `request.cookie_rotations.signed.each do …` and likewise `request.cookie_rotations.encrypted.each do …` (don't think we'd even need the `wrap` then.

mikeycgto replied:
Going to get things integrated using this new configuration API. Will take a little bit of time cause I need to update some tests then 😃

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -54,7 +54,19 @@ module ActiveSupport
+  # the constructor. This method can be called multiple times and new encryptor
+  # instances will be added to the rotation stack on each call.
+  #
+  #   crypt.rotate old_aead_key, cipher: "aes-256-gcm"
```
`rotate raw_key: old_aead_key…` same for the sample below.

kaspth reviewed (commented):
Think this is coming together really nicely!

mikeycgto commented:
I like this patch a lot. The `on_rotation` seems like the best approach so far. I also think its easier to understand API-wise as well. Good call on moving the MessageVerifier changes to `verified` instead of `verify`. Definitely belongs on that method and not `verify`.

One small note, I had to add the following to both the `verified` and `decrypt_and_verify`:

```diff
diff --git a/activesupport/lib/active_support/messages/rotator.rb b/activesupport/lib/active_support/messages/rotator.rb
index 14e9ed0..1379d58 100644
--- a/activesupport/lib/active_support/messages/rotator.rb
+++ b/activesupport/lib/active_support/messages/rotator.rb
@@ -18,7 +18,7 @@ module Encryptor
 
         AES_CBC_EXPRESSION = /\Aaes-\d+-cbc\z/i.freeze
 
-        def decrypt_and_verify(*args, on_rotation: nil)
+        def decrypt_and_verify(*args, on_rotation: nil, **)
           super
         rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature
           run_rotations(on_rotation) { |encryptor| encryptor.decrypt_and_verify(*args) } || raise
@@ -50,7 +50,7 @@ def extract_signing_key(cipher:, signed_salt: nil, key_generator: nil, secret: n
       module Verifier
         include Rotator
 
-        def verified(*args, on_rotation: nil)
+        def verified(*args, on_rotation: nil, **)
           if verified = super
             verified
           else
```

Without the keyword splatting, lots of `MessageEncryptor`/`MessageVerifier` tests were raising `ArgumentError`s with "unknown keyword: purpose"

mikeycgto commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -590,77 +600,56 @@ def commit(options)
+    class EncryptedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       include SerializedCookieJars
 
+      AES_CBC_CIPHER_REGEX = /\Aaes-\d+-cbc\z/i
```
This was moved to `Messages::Rotator`. I never loved using a regex for this. Do you think I should convert this to an Array (or even a `Set`) of all the possible AES CBC ciphers, which is just:

```
> OpenSSL::Cipher.ciphers.grep(/\Aaes-\d+-cbc\z/i)
 => ["AES-128-CBC", "AES-192-CBC", "AES-256-CBC", "aes-128-cbc", "aes-192-cbc", "aes-256-cbc"]
```

kaspth replied:
We could just simplify it to `/-cbc\z/i` (at which point it's clear enough to just inline in the method) or `cipher.downcase.end_with?("cbc")`

mikeycgto commented on activesupport/lib/active_support/messages/rotator.rb:
```diff
@@ -0,0 +1,87 @@
+            self.class.new(raw_key || extract_key(options), digest: digest, serializer: serializer)
+          end
+
+          def extract_key(key_generator: nil, secret: nil, salt:)
```
See my patch and diff below. Keywords from the metadata need to be pass through these methods.

kaspth replied:
If the keywords from the metadata (`:purpose` I assume?) need to pass into this method, why isn't there a capturing `**` at the end?

If we don't need it, great! I just wanted to make sure it wasn't forgotten 😊

mikeycgto commented on guides/source/configuring.md:
```diff
@@ -487,6 +487,20 @@ Defaults to `'signed cookie'`.
   cookie'`.
 
+* `config.action_dispatch.encrypted_cookie_cipher` sets the cipher to be
+  used for encrypted cookies. This defaults to `"aes-256-gcm"`.
```
I think it's good to let this be configurable. I have this end goal of getting Rails to work with chacha20 support for those who want to get more hands on with OpenSSL deps.

kaspth replied:
👍

mikeycgto commented:
Once we're sure this is the configuration API we want, I can update the security and configuration guides as needed. Just want us to be sure this is the API before I do that :smiley:

mikeycgto commented:
This latest merge reminded me of the following code we're now removing in the cookies middleware. Should this check for a missing `secret_key_base` exist elsewhere in the framework?

```diff
-       if ActiveSupport::LegacyKeyGenerator === key_generator
-          raise "You didn't set secret_key_base, which is required for this cookie jar. " \
-            "Read the upgrade documentation to learn more about this new config option."
```

kaspth commented:
> One small note, I had to add the following to both the `verified` and `decrypt_and_verify`

Huh, strange. I didn't have to add that when I tried the diff and ran some of the action_dispatch tests. But perhaps I was missing the Active Support tests.

Remember that it can't just be `**` it has to be `**options` so we can pass them onto `verifier.verified(*args, **options)` etc.

kaspth commented:
> Should this check for a missing `secret_key_base` exist elsewhere in the framework?

Yup, and it does! After the credentials thing, we're now validating `Rails.application.secret_key_base` when it's called with this: https://github.com/rails/rails/blob/31a8e7c6264b76c3966ee3b0dcd14230ef390fa5/railties/lib/rails/application.rb#L530-L538

kaspth commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -21,12 +21,17 @@ class Railtie < Rails::Railtie # :nodoc:
       "X-Content-Type-Options" => "nosniff"
     }
 
+    config.action_dispatch.cookies_rotations = ActiveSupport::Messages::Configuration.new.cookies_rotations
```
Looks like we're always calling this with `new.cookies_rotations`, then there's no point to the namespace handling in the `Configuration` class and we should just move up `Rotation` to `RotationConfiguration`.

mikeycgto replied:
Should I convert this class to a Singleton? Or should I make `rotate` as well as the `signed` and `encrypted` arrays into class/module methods?

kaspth replied:
Nope, let's keep it a regular class. Assigning it to `Rails.application.config` makes it a "singleton", i.e. accessible across the app.

kaspth commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -21,12 +21,17 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_dispatch.perform_deep_munge = true
 
+    config.action_dispatch.legacy_encryptor_schemes = []
+    config.action_dispatch.legacy_verifier_schemes  = []
```
These now need to be yanked.

kaspth commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -39,7 +44,7 @@ class Railtie < Rails::Railtie # :nodoc:
       ActionDispatch::ExceptionWrapper.rescue_templates.merge!(config.action_dispatch.rescue_templates)
 
-      config.action_dispatch.authenticated_encrypted_cookie_salt = "authenticated encrypted cookie" if config.action_dispatch.use_authenticated_cookie_encryption
+      config.action_dispatch.authenticated_encrypted_cookie_salt ||= "authenticated encrypted cookie"
```
Let's move this to the alongside the other `_salt`s up top and skip the `||=` for `=`.

kaspth commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -21,12 +21,17 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_dispatch.use_authenticated_cookie_encryption = false
```
There's a part of me that thinks we should turn:

```ruby
config.action_dispatch.signed_cookie_salt = "signed cookie"
config.action_dispatch.encrypted_cookie_salt = "encrypted cookie"
config.action_dispatch.encrypted_signed_cookie_salt = "signed encrypted cookie"
config.action_dispatch.use_authenticated_cookie_encryption = false
```

into:

```ruby
config.action_dispatch.cookies = ActiveSupport::OrderedOptions.new
config.action_dispatch.cookies.signed_salt = "signed cookie"
config.action_dispatch.cookies.encrypted_salt = "encrypted cookie"
config.action_dispatch.cookies.encrypted_signed_salt = "signed encrypted cookie"
# These three are old 👆, but these three haven't shipped 👇.
config.action_dispatch.cookies.authenticated_encrypted_salt = "authenticated encrypted cookie"
config.action_dispatch.cookies.authenticated_encryption = false
config.action_dispatch.cookies.rotations = ActiveSupport::Messages::RotationConfiguration.new
```

We'd need to deprecate the old ones. Though I think it's highly unlikely that many apps have overriden those default salts, so the hurt isn't too huge. Either that or we just reassign for something so trivial.
It's probably out of scope for this, so let me look into that separately.

Though by the way, why do we need the `authenticated_encrypted_cookie_salt`? We could just as well reuse the `encrypted_signed_cookie_salt`?

mikeycgto replied:
When I added AEAD support I wanted to use a different salt value as this is a best practice when deriving keys for different uses.

kaspth commented:
Yup, let's move forward with that configuration API :)

mikeycgto commented:
> Remember that it can't just be ** it has to be **options so we can pass them onto verifier.verified(*args, **options) etc.

Good call on this. I can update those calls and even add a test case for it.

mikeycgto commented:
> Yup, and it does! After the credentials thing, we're now validating Rails.application.secret_key_base when it's called with this:

Nice to "centralized" method for validating the configuration!

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -54,7 +54,27 @@ module ActiveSupport
+  #
+  # This class also defines a +rotate+ method which can be used to rotate out
+  # no longer used encryption keys. This method accepts the same arguments as
+  # the constructor. This method can be called multiple times and new encryptor
```
`constructor` reads like a JavaScript word, I'd say `the same arguments as +new+`.

kaspth commented:
Looks like the documentation update is missing and then we need to do some commit squashing. Perhaps the Active Support changes and then the cookie changes are two separate commits? Unless you can spot a division that's more git history apt 😊

mikeycgto commented on activesupport/lib/active_support/messages.rb:
```diff
@@ -0,0 +1,9 @@
+
+    autoload :Configuration
+  end
+end
```
Had a feeling it would be. I can just remove this and `require` the class as needed.

mikeycgto commented:
@kaspth yup, the docs is next up for me. Sounds good on the squash commit. That division makes the most sense to me.

mikeycgto commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -39,7 +42,7 @@ class Railtie < Rails::Railtie # :nodoc:
       ActionDispatch::ExceptionWrapper.rescue_templates.merge!(config.action_dispatch.rescue_templates)
 
-      config.action_dispatch.authenticated_encrypted_cookie_salt = "authenticated encrypted cookie" if config.action_dispatch.use_authenticated_cookie_encryption
+      config.action_dispatch.authenticated_encrypted_cookie_salt ||= "authenticated encrypted cookie" if config.action_dispatch.use_authenticated_cookie_encryption
```
Now I remember why we had the guard here. Setting this salt controls whether or not AEAD is used in the middleware. Specifically, the `upgrade_legacy_hmac_aes_cbc_cookies?` needs this salt not to be set if AEAD is not to be used: https://github.com/mikeycgto/rails/blob/8857c2e242938f4bc43365bc50cbc67d43794dab/actionpack/lib/action_dispatch/middleware/cookies.rb#L261-L266

Perhaps we should update the middleware to use `action_dispatch.use_authenticated_cookie_encryption` directly for this?

kaspth replied:
Yes, the middleware should use that option.

mikeycgto commented:
Thought some more about the `RotationConfiguration` interface and I think we should now allow `:encrypted` and `:signed` rotations to be defined at the same time. The options differ slightly and I think we might run into some problems as a result.

Additionally, allowing both encrypted and signed cookies to have the same configuration would lead to the same salt and thus the same key being used for different security features which is not a good idea security-wise. Thoughts @kaspth? Should be easy to remove this small part of the class.

mikeycgto commented:
Updated the guides and the documentation some more.

I hope it's OK that I reworked the session cookie section of the Security guide a bit. Just felt odd at this point to have that section lead with legacy details. I'd rather explain security for the current version upfront then explain differences in prior versions after.

Want to just review it again with fresh eyes and then squash the commits.

kaspth commented:
Let's use the `action_dispatch.use_authenticated_cookie_encryption` option in the middleware then squash the commits.

I'm going to be out of the country the week after this. So this has to be wrapped up soon. If you can get to the squashing tomorrow then great. Otherwise I'll merge and do some follow up codewise then 😊

mikeycgto commented:
Now I'm actually a bit confused on how `action_dispatch.use_authenticated_cookie_encryption` should work. Setting this value to `false` should lead to CBC cookies being used, correct?

For Rails 5.2 this value will default to `true` and CBC cookie upgrading will be added as a rotation (because the CBC salts are still set and `upgrade_legacy_hmac_aes_cbc_cookies?` is `true`).

I wrote a test case for `use_authenticated_cookie_encryption` to false and it's failing. I think somewhere between this PR and the AEAD cookie PR the EncryptedCookieJar become to simplified. Should be easy to fix this though but I want to make sure I'm clear on what the role of `use_authenticated_cookie_encryption` exactly is.

mjc-gh committed:
Add key rotation message Encryptor and Verifier

Both classes now have a rotate method where new instances are added for
each call. When decryption or verification fails the next rotation
instance is tried.

mjc-gh committed:
Add key rotation cookies middleware

Using the action_dispatch.cookies_rotations interface, key rotation is
now possible with cookies. Thus the secret_key_base as well as salts,
ciphers, and digests, can be rotated without expiring sessions.

mikeycgto commented:
Squashed the commits. I think I broke something with the `use_authenticated_cookie_encryption` change. Will try to fix that ASAP.

kaspth commented:
Great! The build looked fine and green to me 😄👏

mikeycgto commented:
One last thing, @kaspth: what about having the cookies middleware use CBC encryption when `use_authenticated_cookie_encryption` is set to `false`? I can add a test case and small patch to the cookies middleware to support this in a separate PR if need be.

kaspth commented:
What's it doing currently? I think `false` and `nil` should probably work the same for that config (depending on which Rails version the app was generated on).

mikeycgto commented:
Currently it's just responsible for enabling the upgrade behavior, which rotates CBC to GCM. I feel setting it to `false` should also lead to CBC being used. The following patch should suffice:

```diff
diff actionpack/lib/
diff --git a/actionpack/lib/action_dispatch/middleware/cookies.rb b/actionpack/lib/action_dispatch/middleware/cookies.rb
index b383164..dbe46be 100644
--- a/actionpack/lib/action_dispatch/middleware/cookies.rb
+++ b/actionpack/lib/action_dispatch/middleware/cookies.rb
@@ -599,9 +599,16 @@ class EncryptedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       def initialize(parent_jar)
         super
 
-        key_len = ActiveSupport::MessageEncryptor.key_len(encrypted_cookie_cipher)
-        secret = request.key_generator.generate_key(request.authenticated_encrypted_cookie_salt, key_len)
-        @encryptor = ActiveSupport::MessageEncryptor.new(secret, cipher: encrypted_cookie_cipher, serializer: SERIALIZER)
+        if request.use_authenticated_cookie_encryption
+          key_len = ActiveSupport::MessageEncryptor.key_len(encrypted_cookie_cipher)
+          secret = request.key_generator.generate_key(request.authenticated_encrypted_cookie_salt, key_len)
+          @encryptor = ActiveSupport::MessageEncryptor.new(secret, cipher: encrypted_cookie_cipher, serializer: SERIALIZER)
+        else
+          key_len = ActiveSupport::MessageEncryptor.key_len("aes-256-cbc")
+          secret = request.key_generator.generate_key(request.encrypted_cookie_salt, key_len)
+          sign_secret = request.key_generator.generate_key(request.encrypted_signed_cookie_salt)
+          @encryptor = ActiveSupport::MessageEncryptor.new(secret, sign_secret, cipher: "aes-256-cbc", serializer: SERIALIZER)
+        end
 
         request.cookies_rotations.encrypted.each do |rotation_options|
           @encryptor.rotate serializer: SERIALIZER, **rotation_options
```

kaspth commented:
Had a bit of a change of heart on the rotate API as I realized that neither the verifier or encryptor ever mentions key generators or salts. They're only meant to work with generated secrets, so I didn't like to add the extra support there after all.

Think we'll have to go through `Rails.application.key_generator` somehow.

The cookies middleware then felt a little strange, so I chose to nix the `:salt` and `:key_generator` support for now. I wasn't happy with how it felt at the moment. Perhaps this is good enough for 5.2, then we can let it evolve a bit on its own once its out.

The updated code is in 36888b9387..38308e6d13

mikeycgto commented:
Sounds good @kaspth. The changes in the documentation look good to me 👍 

Let me know if I should submit my `use_authenticated_cookie_encryption`  PR. I feel like the expectation would be if this is `false`, CBC should be used for cookies.
